client/auth/login.tsx

```diff
@@ -0,0 +1,241 @@
+import queryString from 'query-string'
+import React, { useEffect, useRef, useState } from 'react'
+import { useTranslation } from 'react-i18next'
+import {
+  PASSWORD_MINLENGTH,
+  USERNAME_MAXLENGTH,
+  USERNAME_MINLENGTH,
+  USERNAME_PATTERN,
+} from '../../common/constants'
+import { useForm } from '../forms/form-hook'
+import SubmitOnEnter from '../forms/submit-on-enter'
+import { composeValidators, maxLength, minLength, regex, required } from '../forms/validators'
+import { RaisedButton } from '../material/button'
+import { push } from '../navigation/routing'
+import LoadingIndicator from '../progress/dots'
+import { useAppDispatch, useAppSelector } from '../redux-hooks'
+import { useStableCallback } from '../state-hooks'
+import { logIn } from './action-creators'
+import {
+  AuthBody,
+  AuthBottomAction,
+  AuthCheckBox,
+  AuthContent,
+  AuthContentContainer,
+  AuthPasswordTextField,
+  AuthTextField,
+  AuthTitle,
+  BottomActionButton,
+  FieldRow,
+  ForgotActionButton,
+  LoadingArea,
+  RowEdge,
+  Spacer,
+} from './auth-content'
+import { redirectIfLoggedIn } from './auth-utils'
+import { UserErrorDisplay } from './user-error-display'
+
+interface LoginModel {
+  username: string
+  password: string
+  rememberMe: boolean
+}
+
+export function Login() {
+  const { t, i18n } = useTranslation()
+  const dispatch = useAppDispatch()
+  const auth = useAppSelector(s => s.auth)
+
+  const [isLoading, setIsLoading] = useState(false)
+  const [lastError, setLastError] = useState<Error>()
+
+  useEffect(() => {
+    redirectIfLoggedIn({ auth })
+  }, [auth])
+
+  const abortControllerRef = useRef<AbortController>()
+
+  const onFormSubmit = useStableCallback((model: LoginModel) => {
+    setIsLoading(true)
+    setLastError(undefined)
+
+    abortControllerRef.current?.abort()
+    abortControllerRef.current = new AbortController()
+
+    dispatch(
+      logIn(
+        {
+          username: model.username,
+          password: model.password,
+          remember: model.rememberMe,
+          locale: i18n.language,
+        },
+        {
+          onSuccess: () => {},
+          onError: err => {
+            setIsLoading(false)
+            setLastError(err)
+          },
+          signal: abortControllerRef.current.signal,
+        },
+      ),
+    )
+  })
+
+  const queryModel: { username?: string } = queryString.parse(window.location.search)
+  const { onSubmit, bindInput, bindCheckable, getInputValue } = useForm<LoginModel>(
+    {
+      username: queryModel.username ?? '',
+      password: '',
+      rememberMe: false,
+    },
+    {
+      username: composeValidators(
+        required(t('auth.usernameValidator.required', 'Enter a username')),
+        minLength(
+          USERNAME_MINLENGTH,
+          t('auth.usernameValidator.minLength', {
```

Comment:
I really don't understand why we have different strings for this and password. They are the same string, they have the same context (they don't reference anything about the field they are on). You can likely use the same string for *every* min length validator in the app (and max length as well)

Reply:
There's actually two versions of the minLength and maxLength strings. One (on the login page) is:
```
Enter at least {{minLength}} characters
```
and the other one (on signup page) is:
```
Use at most {{maxLength}} characters
```
I guess the slight difference in verbiage makes sense given the context. I named the keys for thsi `minLength` and `minLength2` fwiw.

But yeah, we can at least use the same keys wherever the string is same.

Reply:
That's a minlength versus a maxlength, so yes, I would expect those strings to be different :) My point was that *all* minlength strings in the app can be the same, and *all* maxlength strings in the app can be the same, not that we need to combine minlength and maxlength.

What you have as "minLength" and "minLength2" are the same thing, they are just for the password and confirmation password, respectively.

Reply:
Oh sorry, I copy-pasted the wrong thing :p

What I meant to indicate is that in some places we use the `Enter ...` (e.g. when logging in), and in some places we use `Use ...` (e.g. when creating a new account). Should we preserve that difference or use the same word in all cases (and which one)?

Reply:
No, that difference is just weirdness in how this stuff got created/written initially, just make all the strings the same.

When you're doing this stuff, consider how we can minimize the amount of work it takes to translate the app, this is the prime concern. If there are very minor content differences that serve no real purpose and provide no actual value to the app, we should be seeking to standardize stuff so it only needs to be translated once.